### PR TITLE
Add json query support for index

### LINKED ISSUE #9: News feed in cards block
Latest news articles must appear in cards bock on the home page ordered by date in descending order.
See https://www.sunstar.com/
image

### PR DESCRIPTION
**Pull request by sdmcraft**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #9 

Seeking early feedback on this (currently DRAFT) PR for adding a generic querying mechanism for index. 
Sunstar site uses article feeds heavily at many places wherein the feeds have filtering criteria (like related, featured, recent, category etc.) where this may be useful. 

Test URLs:
- Before: https://main--sunstar--hlxsites.hlx.page/
- After: https://i-9--sunstar--hlxsites.hlx.page/

- [ ]  If you are adding a new block or a variation to an existing block, has it been added in the [block-library](https://adobe.sharepoint.com/:x:/r/sites/HelixProjects/Shared%20Documents/sites/sunstar/sunstar/sidekick/library.xlsx?d=w7f8a415fe192444caa87806b565bb62d&csf=1&web=1&e=oJWCS0) ?

### CODE REVIEW

**Comment by bosschaert** [comment]
Looks useful to me. I wonder if there is an existing lightweight JS library that can maybe perform the query on the JSON?

Although I must admit I looked around briefly but couldn't find anything as lightweight as your json-query.js 😄

**Comment by sdmcraft** [comment]
> Looks useful to me. I wonder if there is an existing lightweight JS library that can maybe perform the query on the JSON?
> 
> Although I must admit I looked around briefly but couldn't find anything as lightweight as your json-query.js 😄

Yes, I tried searching around and couldn't find as specific as what I was looking for.

File: scripts/json-query.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/* eslint-disable max-len */
+export default function query(select, from, where, orderBy, rowCount) {
```
Comment (jindaliiita) [inline scripts/json-query.js:2]: it would be good if we can include some test for this

File: scripts/json-query.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  // Validate input parameters
+  if (!Array.isArray(select) || !Array.isArray(from) || !Array.isArray(where) || where.length !== 3) {
```
Comment (jindaliiita) [inline scripts/json-query.js:4]: can we replace this with !([select, from, where].every(Array.isArray) && where.length === 3)

File: scripts/scripts.js [hunk 3]
```diff
@@ -332,6 +334,10 @@ export async function fetchIndex(indexFile, sheet, pageSize = 500) {
   return newIndex;
 }
 
+export function queryIndex(index, select, where, orderBy, rowCount) {
```
Comment (jindaliiita) [inline scripts/scripts.js:337]: why wrapper function is required here ?

File: scripts/json-query.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  // Extract the filter condition components
+  const [field, operator, value] = where;
+
+  // Sort the 'from' array based on the orderBy parameter
+  if (orderBy) {
+    const [orderByField, sortOrder] = orderBy;
+    from.sort((a, b) => {
+      const aValue = a[orderByField];
+      const bValue = b[orderByField];
+      if (sortOrder === 'asc') {
```
Comment (jindaliiita) [inline scripts/json-query.js:17]: Conditions can be simplified

```suggestion
      if (sortOrder === 'asc') {
  Math.sign(aValue - bValue)
} else if (sortOrder === 'desc') {
   Math.sign(bValue - aValue)
}
```

File: scripts/json-query.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        return item[field] > parseFloat(value);
+      case '=':
+        return item[field] === value;
+      default:
+        return false; // Unsupported operator
+    }
+  });
+
+  // Limit the number of rows based on the rowCount parameter
+  const selectedData = filteredData.slice(0, rowCount);
```
Comment (jindaliiita) [inline scripts/json-query.js:43]: Can we add pagination support also in this ?

**Comment by sdmcraft** [comment]
Thanks for reviewing. I found a [library (jslinq)](https://github.com/maurobussini/jslinq) which seems to do what we want, providing a richer api and is quite lightweight. Hence closing this PR and opening a new one #63 for it.

**Pull request closed without merging**